File: cpp/bc2026/src/bestsofar/Utils/Movement.cpp
#include "Utils/Movement.hpp"

#include "Utils/Globals.hpp"
#include "Utils/Navigator.hpp"
#include "Utils/Vision.hpp"

namespace Utils {

MapLocation Movement::targetLoc = MapLocation::NONE;
MapLocation Movement::explorePoints[144] = {};
int Movement::epSize = 0;
int Movement::turnsStuck = 0;
int Movement::bestDist = 0;

void Movement::indicateAllExplorePoints() {
    for (const MapLocation& loc : explorePoints) {
        rc.setIndicatorDot(loc, 100, 255, 255);
    }
}

void Movement::setNextExplorePoint() {
    targetLoc = explorePoints[rng.nextInt(epSize)];
    turnsStuck = 0;
    bestDist = rc.getLocation().distanceSquaredTo(targetLoc);
}

void Movement::explore() {
    // if theres no target, or reached target, or see target but unreachable
    if (targetLoc.isNull() || rc.getLocation() == targetLoc || turnsStuck > 14 ||
        (Vision::hasSeenLocation(targetLoc) && !Vision::sensePassability(targetLoc))) {
        setNextExplorePoint();
    }

    Navigator::moveTo(targetLoc, true);

    int newDist = rc.getLocation().distanceSquaredTo(targetLoc);
    if (newDist < bestDist) {
        bestDist = newDist;
        turnsStuck = 0;
    } else turnsStuck++;
}

void Movement::init() {
    // Java: explorePoints = new MapLocation[144] (a fresh, all-null array)
    for (int i = 0; i < 144; i++) explorePoints[i] = MapLocation::NONE;

    for (int i = mapWidth / 15; i < mapWidth / 2; i += (i + 6) / 2) {
        for (int j = mapHeight / 15; j < mapHeight / 2; j += (j + 6) / 2) {
            explorePoints[epSize++] = MapLocation(i, j);
            explorePoints[epSize++] = MapLocation(mapWidth - i - 1, j);
            explorePoints[epSize++] = MapLocation(i, mapHeight - j - 1);
            explorePoints[epSize++] = MapLocation(mapWidth - i - 1, mapHeight - j - 1);
        }
    }
}

void Movement::exploreSafe() {
}

void Movement::circle(MapLocation loc) {
    if (!rc.isMovementReady()) return;
    MapLocation target = loc.add(rc.getLocation().directionTo(loc).rotateLeft());
    if (!rc.onTheMap(target)) target = loc;
    Navigator::moveTo(target, false);
}

}  // namespace Utils

File: py/bc2026/src/bestsofar/Utils/Movement.py
# Port of bestsofar/Utils/Movement.java. See py/CONVENTIONS.md.
from api import *
from Utils.Globals import Globals
from Utils.Navigator import Navigator
from Utils.Vision import Vision


class Movement(Globals):

    targetLoc = None
    explorePoints = None  # MapLocation[]
    epSize = 0
    turnsStuck = 0
    bestDist = 0

    @staticmethod
    def indicateAllExplorePoints():
        for loc in Movement.explorePoints:
            Globals.rc.setIndicatorDot(loc, 100, 255, 255)

    @staticmethod
    def setNextExplorePoint():
        Movement.targetLoc = Movement.explorePoints[Globals.rng.nextInt(Movement.epSize)]
        Movement.turnsStuck = 0
        Movement.bestDist = Globals.rc.getLocation().distanceSquaredTo(Movement.targetLoc)

    @staticmethod
    def explore():
        # if theres no target, or reached target, or see target but unreachable
        if Movement.targetLoc is None or Globals.rc.getLocation() == Movement.targetLoc or Movement.turnsStuck > 14 or (Vision.hasSeenLocation(Movement.targetLoc) and not Vision.sensePassability(Movement.targetLoc)):
            Movement.setNextExplorePoint()

        Navigator.moveTo(Movement.targetLoc, True)

        newDist = Globals.rc.getLocation().distanceSquaredTo(Movement.targetLoc)
        if newDist < Movement.bestDist:
            Movement.bestDist = newDist
            Movement.turnsStuck = 0
        else:
            Movement.turnsStuck += 1

    @staticmethod
    def init():
        Movement.explorePoints = [None] * 144

        # Java: for (int i = mapWidth/15; i < mapWidth/2; i += (i+6)/2)
        # (non-constant step -> while; all operands non-negative so // is exact)
        i = Globals.mapWidth // 15
        while i < Globals.mapWidth // 2:
            j = Globals.mapHeight // 15
            while j < Globals.mapHeight // 2:
                Movement.explorePoints[Movement.epSize] = MapLocation(i, j)  # Java: explorePoints[epSize++]
                Movement.epSize += 1
                Movement.explorePoints[Movement.epSize] = MapLocation(Globals.mapWidth - i - 1, j)
                Movement.epSize += 1
                Movement.explorePoints[Movement.epSize] = MapLocation(i, Globals.mapHeight - j - 1)
                Movement.epSize += 1
                Movement.explorePoints[Movement.epSize] = MapLocation(Globals.mapWidth - i - 1, Globals.mapHeight - j - 1)
                Movement.epSize += 1
                j += (j + 6) // 2
            i += (i + 6) // 2

    @staticmethod
    def exploreSafe():
        pass

    @staticmethod
    def circle(loc):
        if not Globals.rc.isMovementReady():
            return
        target = loc.add(Globals.rc.getLocation().directionTo(loc).rotateLeft())
        if not Globals.rc.onTheMap(target):
            target = loc
        Navigator.moveTo(target, False)
